jupyterlab_chameleon/artifact.py

```diff
@@ -1,66 +1,257 @@
-import tarfile
-
-from dataclasses import asdict
 import json
+import logging
 import os
+import requests
+import tarfile
 import tempfile
-import zipfile
-
-from keystoneauth1.exceptions.http import Unauthorized
+from dataclasses import asdict
 from jupyter_server.base.handlers import APIHandler
-import requests
+from keystoneauth1.exceptions.http import Unauthorized
+from requests import HTTPError
 from tornado import web
-from traitlets import Any, CRegExp, Int, Unicode
+from traitlets import Any, CRegExp, Int
 from traitlets.config import LoggingConfigurable
 
-from .db import Artifact, DB
-from .exception import AuthenticationError, BadRequestError, IllegalArchiveError
-from .util import call_jupyterhub_api, ErrorResponder
+from .db import DB
+from .exception import AuthenticationError, IllegalArchiveError, BadRequestError
+from .trovi import contents_url, get_trovi_token, artifacts_url, artifact_versions_url
+from .util import ErrorResponder
 
-import logging
 LOG = logging.getLogger(__name__)
 
 
+class ArtifactAuthor:
+    def __init__(self, full_name, email, affiliation=None):
+        """
+        An author of an artifact
+
+        Attrs:
+            full_name (str): the full name of the author
+            affiliation (str): the institution/organization with which the author
+                is affiliated
+            email (str): the author's email address
+        """
+        self.full_name = full_name
+        self.affiliation = affiliation
+        self.email = email
+
+
+class ArtifactVersion:
+    def __init__(
+        self, contents_location, contents_id, links=None, created_at=None, slug=None
+    ):
+        """
+        A version of an artifact
+
+        Attrs:
+            contents_location (str): represents where the contents are stored
+            contents_id (str): an ID which is used to look up the contents at the
+                storage location
+            links (list[str]): a URN which represents a link relevant to the artifact
+            created_at (datetime): the time at which this version was created
+            slug (str): the slug for this version's URL
+        """
+        self.contents_location = contents_location
+        self.contents_id = contents_id
+        self.links = links or []
+        self.created_at = created_at
+        self.slug = slug
+
+    @property
+    def contents_urn(self):
+        return f"urn:{self.contents_location}:{self.contents_id}"
+
+
+class Artifact:
+    """A shared experiment/research artifact that can be spawned in JupyterHub.
```

Comment:
Docstring doesn't line up w/ implementation. TBH I don't think you really need documentation here, it's not a library, rather it is adapting a Trovi API object, if anything you could link to Trovi docs?

That said, you already have them, and they're useful for code inspection 👍🏻